core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala

```diff
@@ -240,4 +240,73 @@ trait AssociativeFlattenSyntax {
     def flatMap[B](f: A => F[B])(implicit flatten: AssociativeFlatten[F], covariant: Covariant[F]): F[B] =
       flatten.flatten(covariant.map(f)(fa))
   }
+
+  /**
+   * Provides infix syntax for filtering covariant types.
+   */
+  implicit class AssociativeFlattenCovariantIdentityBothIdentityEitherOps[F[+_], A](fa: F[A]) {
+
+    /**
+     * Collects values of `A` for which the partial function `pf` is defined.
+     */
+    def collect[B](pf: PartialFunction[A, B])(implicit
+      flatten: AssociativeFlatten[F],
+      covariant: Covariant[F],
+      identityBoth: IdentityBoth[F],
+      identityEither: IdentityEither[F]
+    ): F[B] =
+      fa.flatMap(a => pf.lift(a).fold[F[B]](identityEither.none)(_.succeed))
+
+    /**
+     * Collects values of `A` for which the partial function `pf` is defined.
+     */
+    def collectM[B](pf: PartialFunction[A, F[B]])(implicit
+      flatten: AssociativeFlatten[F],
+      covariant: Covariant[F],
+      identityBoth: IdentityBoth[F],
+      identityEither: IdentityEither[F]
+    ): F[B] =
+      fa.flatMap(a => pf.lift(a).getOrElse(identityEither.none))
+
+    /**
+     * Filters an `F[A]` value with a predicate `f`.
+     */
+    def filter[B](f: A => Boolean)(implicit
+      flatten: AssociativeFlatten[F],
+      covariant: Covariant[F],
+      identityBoth: IdentityBoth[F],
+      identityEither: IdentityEither[F]
+    ): F[A] =
+      fa.flatMap(a => if (f(a)) a.succeed else identityEither.none)
+
+    /**
+     * Filters an `F[A]` value with a predicate `f`.
+     */
+    def filterM[B](f: A => F[Boolean])(implicit
+      flatten: AssociativeFlatten[F],
+      covariant: Covariant[F],
+      identityBoth: IdentityBoth[F],
+      identityEither: IdentityEither[F]
+    ): F[A] =
+      fa.flatMap(a => f(a).flatMap(b => if (b) a.succeed else identityEither.none))
+
+    /**
+     * Filters an `F[A]` value with a predicate `f`.
+     */
+    def filterNot[B](f: A => Boolean)(implicit
```

Comment:
Is it really worth having?

Reply:
It is definitely not fundamental. I included it for consistently with the Scala standard library but we could take it out.

Reply:
I don't mind keeping it. But I have to admit I have always wondered about the value of these `*Not` variants even in the standard library.

Reply:
I took it out.